File: tests/testthat/test-sectors.R
library(testthat)
library(Juliora)

if (!Juliora::is_julia_available()) {
  skip("Julia environment not available for testing")
}

make_sector_test_mrio <- function(indices) {
  n <- nrow(indices)
  transactions <- MatrixEntry(matrix(0, nrow = n, ncol = n), indices, indices)
  final_demand <- MatrixEntry(
    matrix(1, nrow = n, ncol = 1),
    data.frame(Category = "Households"),
    indices
  )
  value_added <- MatrixEntry(
    matrix(1, nrow = 1, ncol = n),
    indices,
    data.frame(Category = "Value added")
  )
  MRIO(transactions, final_demand, value_added)
}

test_that("sectors returns the same character representation for Gloria and Eora", {
  gloria_indices <- data.frame(
    CountryCode = rep(c("AUT", "DEU"), each = 2),
    Sector = rep(c("Agriculture", "Manufacturing"), 2)
  )
  eora_indices <- data.frame(
    CountryCode = rep(c("AUT", "DEU"), each = 2),
    Industry = rep(c("01", "02"), 2),
    Sector = rep(c("Agriculture", "Manufacturing"), 2)
  )

  gloria <- make_sector_test_mrio(gloria_indices)
  eora <- make_sector_test_mrio(eora_indices)
  expected <- c("Agriculture", "Manufacturing")

  expect_type(sectors(gloria), "character")
  expect_identical(sectors(gloria), expected)
  expect_identical(sectors(eora), expected)
  expect_identical(sectors(gloria), sectors(eora))
  expect_identical(sector(gloria), expected)
})

test_that("sectors reports invalid inputs as native R errors", {
  expect_error(sectors(42), "Julia Error")
})

File: test/test_sectors.jl
function sector_test_mrio(indices::DataFrame)
    n = nrow(indices)
    transactions = MatrixEntry(zeros(n, n), indices, indices)
    final_demand_indices = DataFrame(Category = ["Households"])
    final_demand = MatrixEntry(ones(n, 1), final_demand_indices, indices)
    value_added_indices = DataFrame(Category = ["Value added"])
    value_added = MatrixEntry(ones(1, n), indices, value_added_indices)
    return MRIO(Z = transactions, Y = final_demand, VA = value_added)
end

@testset "Eora and Gloria use the same sector representation" begin
    gloria_indices = DataFrame(
        CountryCode = ["AUT", "AUT", "DEU", "DEU"],
        Sector = ["Agriculture", "Manufacturing", "Agriculture", "Manufacturing"],
    )
    eora_indices = DataFrame(
        CountryCode = ["AUT", "AUT", "DEU", "DEU"],
        Industry = ["01", "02", "01", "02"],
        Sector = ["Agriculture", "Manufacturing", "Agriculture", "Manufacturing"],
    )

    gloria = sector_test_mrio(gloria_indices)
    eora = sector_test_mrio(eora_indices)

    expected = ["Agriculture", "Manufacturing"]
    @test (@inferred sectors(gloria)) == expected
    @test (@inferred sectors(eora)) == expected
    @test sectors(gloria) == sectors(eora)
    @test eltype(sectors(gloria)) === String
    @test eltype(sectors(eora)) === String
    @test sector(gloria) == expected
end

@testset "sectors preserves order and normalizes metadata to strings" begin
    indices = DataFrame(Sector = Any["Services", missing, :Agriculture, "Services"])

    @test sectors(indices) == ["Services", "Agriculture"]
    @test sectors(DataFrame(Industry = ["A", "B", "A"])) == ["A", "B"]
    @test sectors(DataFrame(CountryCode = ["AUT"])) == String[]
end
